.NET: migration from SharpRaven (#3401)
getsentry/sentry-docs: src/platforms/dotnet/migration.mdx

#### ASP.NET Core Environment Casing

To match Sentry's default environment naming convention, unless explicitly set otherwise, the Sentry environment will be reported lowercase. That means if `ASPNET_ENVIRONMENT` is equal to `Production`, the Sentry SDK will report it as `production`.

#### ASP.NET Core `IncludeRequestPayload`

The option `IncludeRequestPayload` that was deprecated in the Sentry SDK 2.0 has been removed.
The replacement is the option `MaxRequestBodySize`.

#### `IHub` and `ISentryClient` received new methods

Sentry's Performance product requires additive API changes to interfaces such as `IHub` and `ISentryClient`.
If you implement your own `IHub`, you'll need to implement two new methods:

`ISentryClient` received: `CaptureTransaction`.
`IHub` received: `StartTransaction` and `GetTraceHeader`.

#### `DiagnosticsLogger` renamed to `DiagnosticLogger`

To align with other SDKs, the option is now named: `DiagnosticLogger`
[Docs PR](https://github.com/getsentry/sentry-docs/pull/2945).

> #### `Dsn` is just a string

Reviewer: ```suggestion
#### `Dsn` as a String
```

`SentryOptions` now take a string for `Dsn`:

Before: `o.Dsn = new Dsn("..");`
After: `o.Dsn = "..";`

> #### `LogEntry` became `Message`

Reviewer: ```suggestion
#### `LogEntry` Became `Message`
```

The property of `SentryEvent` that supports a structured log entry was renamed to `Message` to align with the protocol and other SDKs.

> #### Sentry On Premise Support

Reviewer: ```suggestion
#### Sentry Self-Hosted Support
```

This version uses the [envelope endpoint](https://develop.sentry.dev/sdk/envelopes/). If you are using an on-premise installation it requires Sentry version >= v20.6.0 to work. If you are using sentry.io nothing will change and no action is needed.

> ### New features and improvements

Reviewer: ```suggestion
### New Features and Improvements
```

- Event queue depth
- Shutdown timeout: If there are events to send, how long to wait until shutdown
- Accept compressed response
- Compress request body
- Breadcrumb level: Minimum log level to store as a breadcrumb
- Event level: Minimum log level to send an event to Sentry
- Disable duplicate event detection
- Disable capture of global unhandled exceptions
- Add event processor
- Add exception processor
- Enable SDK debug mode
- Attach stack trace for captured messages (opt-in)

### Breaking changes

#### .NET Framework support reduced

If your project is targeting .NET Framework, it will need to be using at least v4.6.1 to use Sentry (SharpRaven supported .NET Framework v3.5 and higher).

#### Initialization

Changed how to initialize Sentry:

```csharp
using Sentry;

// Initialize with DSN
SentrySdk.Init("___PUBLIC_DSN___");

// Initialize with options
SentrySdk.Init(o =>
{
    o.Dsn = "___PUBLIC_DSN___";
});
```

For ASP.NET (classic) integration, add `Sentry.AspNet` package and initialize it by calling `o.AddAspNet()`:

```csharp
using Sentry;
using Sentry.AspNet;

SentrySdk.Init(o =>
{
    o.Dsn = "___PUBLIC_DSN___";

    o.AddAspNet();
});
```

> #### Event queue

Reviewer: ```suggestion
#### Event Queue
```

Contrary to SharpRaven, events in Sentry SDK are sent in background without blocking the currently executing code. Calling `SentrySdk.CaptureException(...)` does not immediately send an event, but instead queues it up on a background thread.

Usage changes:

```csharp
// Before
ravenClient.Capture(...);
await ravenClient.CaptureAsync(...);

// After
SentrySdk.CaptureEvent(...);
```

You can flush the event queue by invoking the `IDisposable` returned by `SentrySdk.Init(...)`:

```csharp
var sentryInitialization = SentrySdk.Init("___PUBLIC_DSN___");

// Capture a non-error message
SentrySdk.CaptureMessage("hello world");

// Tear down Sentry and flush queued up events
sentryInitialization.Dispose();
```

> #### Event processors

Reviewer: ```suggestion
#### Event Processors
```

### Breaking Changes

Upgrading from SharpRaven to Sentry SDK includes the following breaking changes.

#### .NET Framework Support Reduced

If your project is targeting .NET Framework, it will need to be using at least v4.6.1 to use Sentry (SharpRaven supported .NET Framework v3.5 and higher).

#### Initialization

Changed how to initialize Sentry:

```csharp
using Sentry;

// Initialize with DSN
SentrySdk.Init("___PUBLIC_DSN___");

// Initialize with options
SentrySdk.Init(o =>
{
    o.Dsn = "___PUBLIC_DSN___";
});
```

For ASP.NET (classic) integration, add `Sentry.AspNet` package and initialize it by calling `o.AddAspNet()`:

```csharp
using Sentry;
using Sentry.AspNet;

SentrySdk.Init(o =>
{
    o.Dsn = "___PUBLIC_DSN___";

    o.AddAspNet();
});
```

#### Event Queue

Contrary to SharpRaven, events in Sentry SDK are sent in background without blocking the currently executing code. Calling `SentrySdk.CaptureException(...)` does not immediately send an event, but instead queues it up on a background thread.

Usage changes:

```csharp
// Before
ravenClient.Capture(...);
await ravenClient.CaptureAsync(...);

// After
SentrySdk.CaptureEvent(...);
```

> 

Reviewer 2: ```suggestion

If you need to stop the execution until events captured are sent to Sentry, you can also flush the queue:

```csharp
SentrySdk.CaptureException(...);
// Only proceed once exception is captured. Waiting up to 2 seconds for it.
await SentrySdk.FlushAsync(TimeSpan.FromSeconds(2));
```

> You can flush the event queue by invoking the `IDisposable` returned by `SentrySdk.Init(...)`:
> 
> ```csharp
> var sentryInitialization = SentrySdk.Init("___PUBLIC_DSN___");
> 
> // Capture a non-error message
> SentrySdk.CaptureMessage("hello world");
> 
> // Tear down Sentry and flush queued up events
> sentryInitialization.Dispose();

Reviewer 2: ````suggestion
You can close the SDK, which will also flush the event queue, by invoking the `IDisposable` returned by `SentrySdk.Init(...)`:

```csharp
var sentryInitialization = SentrySdk.Init("___PUBLIC_DSN___");

// Capture a non-error message
SentrySdk.CaptureMessage("hello world");

// Tear down Sentry and flush queued up events
sentryInitialization.Dispose();
````

#### Initialization

Changed how to initialize Sentry:

```csharp
using Sentry;

// Initialize with DSN
SentrySdk.Init("___PUBLIC_DSN___");

// Initialize with options
SentrySdk.Init(o =>
{
    o.Dsn = "___PUBLIC_DSN___";
});
```

For ASP.NET (classic) integration, add `Sentry.AspNet` package and initialize it by calling `o.AddAspNet()`:

```csharp
using Sentry;
using Sentry.AspNet;

SentrySdk.Init(o =>
{
    o.Dsn = "___PUBLIC_DSN___";

    o.AddAspNet();
});
```

#### Event Queue

Contrary to SharpRaven, events in Sentry SDK are sent in background without blocking the currently executing code. Calling `SentrySdk.CaptureException(...)` does not immediately send an event, but instead queues it up on a background thread.

Usage changes:

```csharp
// Before
ravenClient.Capture(...);
await ravenClient.CaptureAsync(...);

// After
SentrySdk.CaptureEvent(...);
```

If you need to stop the execution until events captured are sent to Sentry, you can also flush the queue:

```csharp
SentrySdk.CaptureException(...);
// Only proceed once exception is captured. Waiting up to 2 seconds for it.
await SentrySdk.FlushAsync(TimeSpan.FromSeconds(2));

> You can close the SDK, which will also flush the event queue, by invoking the `IDisposable` returned by `SentrySdk.Init(...)`:

Reviewer 2: ````suggestion
```
You can close the SDK, which will also flush the event queue, by invoking the `IDisposable` returned by `SentrySdk.Init(...)`:
````

# Release notes

Please refer to [GitHub for the complete release notes](https://github.com/getsentry/sentry-dotnet/releases/tag/3.0.0).

> <Note>

Reviewer: Looking at the version you're updating, I can see you're following that model by adding a note format here, but since we now have two migration guides (updating to version 3.0.5 and updating from SharpRaven), please make both of those heading level 2 so users can see that there are two migration guides on the page.

Migrating from SharpRaven to Sentry SDK

</Note>

> # Breaking changes

Reviewer: Two things here (please make these changes throughout):
1. Headings use Title Case
2. Headings need to be followed by at least one line of text (or the content for "On this page" doesn't work properly)

```suggestion
# Breaking Changes

Upgrading from SharpRaven to Sentry SDK includes the following breaking changes.
```

Author: > Headings use Title Case

Only h1 or all of them?

Reviewer: All (and I know some content isn't yet updated to this standard, so I'm glad you asked).

SentrySdk.Init(o =>
{
    o.Dsn = "___PUBLIC_DSN___";

    o.AddAspNet();
});
```

## Event queue

Contrary to SharpRaven, events in Sentry SDK are sent in background without blocking the currently executing code. Calling `SentrySdk.CaptureException(...)` does not immediately send an event, but instead queues it up on a background thread.

Usage changes:

```csharp
// Before
ravenClient.Capture(...);
await ravenClient.CaptureAsync(...);

// After
SentrySdk.CaptureEvent(...);
```

You can flush the event queue by invoking the `IDisposable` returned by `SentrySdk.Init(...)`:

```csharp
var sentryInitialization = SentrySdk.Init("___PUBLIC_DSN___");

// Capture a non-error message
SentrySdk.CaptureMessage("hello world");

// Tear down Sentry and flush queued up events
sentryInitialization.Dispose();
```

## Event processors

Usage changes:

```csharp
// Before
ravenClient.BeforeSend = requester =>
{
    // Here you can log data from the requester
    // or replace it entirely if you want.
    return requester;
};

// After
SentrySdk.Init(o =>
{
    o.BeforeSend = e =>
    {
        // Mutate the event or return null to drop it altogether
        return e;
    };
}
});
```

> ## List of differences between SharpRaven and SentrySdk

Reviewer: This list is super useful - I suggest it'd be better to list it prior to the breaking changes to help folks understand the value of migrating.
